hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/PhysicalOpBindJoinWithVALUESorFILTER.java

```diff
@@ -0,0 +1,89 @@
+package se.liu.ida.hefquin.engine.queryplan.physical.impl;
+
+import se.liu.ida.hefquin.base.query.TriplePattern;
+import se.liu.ida.hefquin.base.queryplan.ExpectedVariables;
+import se.liu.ida.hefquin.engine.federation.FederationMember;
+import se.liu.ida.hefquin.engine.federation.SPARQLEndpoint;
+import se.liu.ida.hefquin.engine.queryplan.executable.UnaryExecutableOp;
+import se.liu.ida.hefquin.engine.queryplan.executable.impl.ops.ExecOpBindJoinSPARQLwithVALUESorFILTER;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpBGPAdd;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpGPAdd;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPAdd;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpTPOptAdd;
+import se.liu.ida.hefquin.engine.queryplan.physical.PhysicalPlanVisitor;
+
+/**
+* A physical operator that implements (a batching version of) the bind
+ * join algorithm. It starts by using a VALUES clause to capture the potential join
+ * partners that are sent to the federation member. If this fails, it uses the bind join
+ * algorithm with a FILTER clause instead.
+
+ * <p>
+ * <b>Algorithm description:</b> For a detailed description of the
+ * actual algorithm associated with this physical operator, refer
+ * to {@link ExecOpBindJoinSPARQLwithVALUESorFILTER}, which provides the
+ * implementation of this algorithm.
+ * </p>
+ */
+public class PhysicalOpBindJoinWithVALUESorFILTER extends BaseForPhysicalOpSingleInputJoin
+{
+	public PhysicalOpBindJoinWithVALUESorFILTER( final LogicalOpTPAdd lop ) {
+		super(lop);
+		assert lop.getFederationMember() instanceof SPARQLEndpoint;
+	}
+	public PhysicalOpBindJoinWithVALUESorFILTER( final LogicalOpBGPAdd lop ) {
+		super(lop);
+
+		assert lop.getFederationMember() instanceof SPARQLEndpoint;
+	}
+
+	public PhysicalOpBindJoinWithVALUESorFILTER( final LogicalOpGPAdd lop ) {
+		super(lop);
+
+		assert lop.getFederationMember() instanceof SPARQLEndpoint;
+	}
+
+	@Override
+	public boolean equals( final Object o ) {
+		return o instanceof PhysicalOpBindJoinWithVALUESorFILTER
+				&& ((PhysicalOpBindJoinWithVALUESorFILTER) o).lop.equals(lop);
+	}
+
+	@Override
+	public UnaryExecutableOp createExecOp( final boolean collectExceptions,
+	                                       final ExpectedVariables... inputVars ) {
+		final TriplePattern pt;
+		final FederationMember fm;
+
+		if ( lop instanceof LogicalOpTPAdd ) {
+			pt = ( (LogicalOpTPAdd) lop ).getTP();
+			fm = ( (LogicalOpTPAdd) lop ).getFederationMember();
+		}
+		else {
```

Comment:
You need to add cases for the other two types of logical operators for which you have added the two extra constructors. See how this is done in `PhysicalOpBindJoinWithFILTER`.

In this context, you will also need to add two more constructors to `ExecOpBindJoinSPARQLwithVALUESorFILTER`.